frame/data-feed/src/lib.rs

```diff
@@ -0,0 +1,721 @@
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			// parse origin
+			ActiveParamTypes::try_mutate::<_, DispatchError, _>(|v| {
+				if v.contains(&key) {
+					Err(Error::<T>::ExistedKey)?;
+				}
+				if v.len() + 1 > MAX_KEY_LEN {
+					Err(Error::<T>::ExceedLimit)?
+				}
+
+				v.push(key.clone());
+				Ok(())
+			})?;
+			Infos::<T>::insert(&key, info);
+			Self::deposit_event(RawEvent::RegisterStorageKey(key));
+			Ok(())
+		}
+
+		/// remove the storage key from the limited set so the data feed pallet no longer
+		/// can change the corresponding value afterwards.
+		#[weight = T::WeightInfo::remove_storage_key(key.len() as u32)]
+		pub fn remove_storage_key(origin, key: StorageKey) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			ActiveParamTypes::mutate(|v| {
+				// remove key
+				v.retain(|k| k != &key);
+			});
+			Infos::<T>::remove(&key);
+			Url::remove(&key);
+			DataFeeds::<T>::remove_prefix(&key);
+			Self::deposit_event(RawEvent::RemoveStorageKey(key));
+			Ok(())
+		}
+
+		/// Set a url for a key which used in offchain to fetch data from this url.
+		#[weight = T::WeightInfo::set_url(key.len() as u32, url.len() as u32)]
+		pub fn set_url(origin, key: StorageKey, url: Vec<u8>) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			Url::insert(&key, &url);
+			Self::deposit_event(RawEvent::SetUrl(key, url));
+			Ok(())
+		}
+
+		/// Submit a new data under the specific storage key.
+		#[weight = T::WeightInfo::feed_data(key.len() as u32)]
+		pub fn feed_data(origin, key: StorageKey, value: DataType) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::feed_data_impl(who, key, value)
+		}
+
+		/// Add a provider to current provider collection.
+		#[weight = T::WeightInfo::add_provider()]
+		pub fn add_provider(origin, new_one: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			let new_one = T::Lookup::lookup(new_one)?;
+
+			ActiveProviders::<T>::try_mutate(|v| -> DispatchResult {
+				if !v.contains(&new_one) {
+					if v.len() + 1 > MAX_KEY_LEN {
+						Err(Error::<T>::ExceedLimit)?
+					}
+
+					v.push(new_one.clone());
+				}
+				Ok(())
+			})?;
+			Self::deposit_event(RawEvent::AddProvider(new_one));
+			Ok(())
+		}
+
+		/// Remove a provider from current provider collection.
+		#[weight = T::WeightInfo::remove_provider()]
+		pub fn remove_provider(origin, who: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			let who = T::Lookup::lookup(who)?;
+
+			ActiveProviders::<T>::mutate(|v| {
+				v.retain(|accountid| accountid != &who);
+			});
+			Self::deposit_event(RawEvent::RemoveProvider(who));
+			Ok(())
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {
```

Comment:
```suggestion
		fn on_finalize(n: T::BlockNumber) {
```

Reply:
Ideally this logic should be done in `on_initialize`. Is there any difference to you?

Reply:
fixed